llpc/context/llpcCompiler.cpp

```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
   }
 
   if (result != Result::Success) {
-    result = buildPipelineInternal(context, shaderInfo, /*unlinked=*/false, pipelineElf, stageCacheAccesses);
+    // See if we want to use part-pipeline compilation, if there is an FS and at least one pre-rasterization
+    // shader stage.
+    unsigned stageMask = context->getShaderStageMask();
+    bool buildPartPipeline = (cl::EnablePartPipeline && (stageMask & (1 << ShaderStageFragment)) &&
+                              (stageMask & ~(1 << ShaderStageFragment)));
+    if (buildPartPipeline)
+      result = buildGraphicsPipelineWithPartPipelines(context, shaderInfo, pipelineElf, stageCacheAccesses);
+    else
+      result = buildPipelineInternal(context, shaderInfo, PipelineLink::WholePipeline, nullptr, pipelineElf, stageCacheAccesses);
   }
   releaseContext(context);
   return result;
 }
 
+// =====================================================================================================================
+// Build graphics pipeline using part-pipeline compilation
+//
+// @param graphicsContext : Graphics context this graphics pipeline
+// @param shaderInfo : Shader info of this graphics pipeline
+// @param [out] pipelineElf : Output Elf package
+// @param [out] stageCacheAccesses : Stage cache access result. All elements must be initialized in the caller as
+//                                   CacheAccessInfo::CacheNotChecked
+Result Compiler::buildGraphicsPipelineWithPartPipelines(Context *context,
+                                                        ArrayRef<const PipelineShaderInfo *> shaderInfo,
+                                                        ElfPackage *pipelineElf,
+                                                        MutableArrayRef<CacheAccessInfo> stageCacheAccesses) {
+  Result result = Result::Success;
+  unsigned wholeStageMask = context->getShaderStageMask();
+  ElfPackage partPipelineBuffers[2];
+  std::unique_ptr<Pipeline> elfLinkerPipeline;
+  std::unique_ptr<ElfLinker> elfLinker;
```

Comment:
It might be clearer if you just defined an ElfLinker in the stack, and then checked which "partIdx" you were compiling to determine what to do with the elf linker.  I'll add other comments below.

Reply:
Reworked this part to check for the stage type instead (didn't go for the stack approach for consistency with other functions).